### ISSUE TITLE
Documentation should take class extensions into consideration

### ISSUE THREAD

**Issue by andreespirela** [denoland/deno_doc#136]
As of right now, `deno doc` does not respect extensions. This means

```
export class B { 
   public id: number;
}
```

```
export class A extends B { 
  public userName: string;
}
```
if I run `deno doc` for class `A`, it will only document the property `userName` but **will not** for `id` part of the class extension.

`deno doc` should be able to read this, and respect the extensions of extensions (example, B extends C)

**Comment by kt3k** [comment]
There was a feedback from std users that inherited properties not being shown in API docs is confusing https://github.com/denoland/deno_std/issues/4353

**Comment by jsejcksn** [comment]
> Can we consider listing all properties including inherited ones?

I think we should not list all properties, but instead hyperlink to the associated symbol (e.g. type alias/interface/class/etc.). This will prevent a significant amount of noise.

Ref: https://github.com/denoland/deno_std/issues/4353#issuecomment-1962927306

**Comment by kt3k** [comment]
Hyperlink is already available in doc website (jsr). For example, see https://jsr.io/@std/csv/doc/~/CsvParseStreamOptions

> This will prevent a significant amount of noise.

I'm not sure I agree these are noises.

If someone comes to the page for `CsvParseStreamOptions`, then they probably want to know the all available options in `CsvParseStreamOptions`. Currently they need to see 2 separate pages to know all options, and that is not convenient for them.

**Comment by jsejcksn** [comment]
> I'm not sure I agree these are noises.
> 
> If someone comes to the page for `CsvParseStreamOptions`, then they probably want to know the all available options in `CsvParseStreamOptions`.

I agree that in cases where the interface is fully bespoke, a reader would certainly want to understand the shape of the type in full (because it is completely unfamiliar).

However, there are many cases where types are derived from other types (e.g. extension, inheritance, intersection, etc.) — and in many cases the existing types are already familiar, so the expanded type information serves no purpose (it's only noise). Here's a contrived example for illustration:

[TS Playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&removeComments=true&target=99&jsx=4&exactOptionalPropertyTypes=true&inlineSourceMap=true&inlineSources=true&isolatedModules=true&noImplicitOverride=true&noErrorTruncation=true#code/PTAEGEEMDtQIwKagK4GcEBNQBcD2oBzBbUAW1wCckBLaAM3zot1NAElpsEAbb6gZQTR0ALgBQ2AJ4AHJAFEAHtJgYAPABUAfKAC8oAN6gA2gGlQtUAGsEk3HVDqAuiIenHoAL4BuMRJlIAGVprDAA5XAwEDW09dVAAMgMxAEgAYwALam4MAH4XIOgQ8Mjon2TlKk480AKiiKitH29fKVlQACUEVGlcYQRipD1azAHVTu7e9E0fPzaACVwANwQKdUzUdVwANWoEAHd1fw4GAAZdUEVlaDVxnr6B6bEQUBeAPRyngCoxfRTICgokEkACFkHQ6CsXAAKACUum0AAVmKRqOhVABBAFA0HglaPZJwbi4ODQuE6RHI1FRYFEuD4qiQDC9biSeARSQuTqMyCEhD8bAM0iqACqtGwAA5MYDJNoAD6gaDIXhlBlM6AstkYSTC9AYFxwXC4bgIGBlDJZXL5YIjepjLp3KZmonQBCk+Ede2TBBlBgUUgAEUg2EgbvJoCRLCpqgAYpQA0HIPSTWqNelkytUC45umKKgygArVC9UMUyNomAylXJ5ms3CWfWG42mlIVITYarDMK225epOMmugKgYahUVJcPVso0m6BV-vq1moYPYNAuRWkRAUWcphdLtDqBAKbAuRcUWgEMpcQ8l8OUtEns997c4fycz19Q6yLcD5AUbjHgVnk0YifMAYhAA)

```ts
type LinkedNode<T> = T & {
  child?: LinkedNode<T>;
  parent?: LinkedNode<T>;
};

type ResponseNode = LinkedNode<Response>;
/*   ^?
{
  arrayBuffer: () => Promise<ArrayBuffer>;
  blob: () => Promise<Blob>;
  readonly body: ReadableStream<Uint8Array> | null;
  readonly bodyUsed: boolean;
  child?: LinkedNode<Response>;
  clone: () => Response;
  formData: () => Promise<FormData>;
  readonly headers: Headers;
  json: () => Promise<any>;
  readonly ok: boolean;
  parent?: LinkedNode<Response>;
  readonly redirected: boolean;
  readonly status: number;
  readonly statusText: string;
  text: () => Promise<string>;
  readonly type: ResponseType;
  readonly url: string;
}
*/
```

In the example above, `Response` is a standard, built-in type, and well-known by many developers. I think it is noise to include all of the fields in `Response`, and — similarly — I think it is noise to include the properties from `LinkedNode` because they both already exist as named types and can be referenced. If the reader is already familiar with either/both of them, then inlining all of the content makes it more difficult to understand which fields are unique to the type in question (there are none in the example, but other types might have them).

> Currently they need to see 2 separate pages to know all options, and that is not convenient for them.

If the user is not familiar with `Response`, then perhaps they're also unfamiliar with `FormData`, which is in the return type of the field `formData`:

```ts
formData: () => Promise<FormData>;
```

It's also inconvenient for the user to click `FormData` to see its type information, right? It seems arbitrary that we would expand and inline all of the type information for the top-level fields, but not any of the type information for nested fields. I'm not sure that the convenience reasoning is scalable.

**Comment by kt3k** [comment]
Ok. The above case looks like a valid case where showing inherited props can be noisy.

Then how about showing some text link saying `show inherited properties` at the bottom of properties? and when a user clicks that button, the inherited props are expanded there. That would cater for both needs, I think

**Comment by jsejcksn** [comment]
> Then how about showing some text link saying `show inherited properties` at the bottom of properties? and when a user clicks that button, the inherited props are expanded there. That would cater for both needs, I think

I think that's a great feature idea!

**Comment by NfNitLoop** [comment]
My current workaround is to do this:

```ts
type Base = { … }

class Impl implements Base {
    /** Implements {@link Base.foo} */
    foo(): void { /* … */ }

    /* … */
}
```

This at least gives people browsing JSR.io direct links to the relevant documentation, and allows me to not have to keep it updated in N places. But it does make the in-IDE experience a bit worse.

[example code](https://github.com/NfNitLoop/better-iterators/commit/201c0f84a91ff705bc0a93781ed1e441a5ad731c) with [view in jsr.io](https://jsr.io/@nfnitloop/better-iterators@2.0.0-rc3/doc/~/Lazy)
